src/hotspot/share/cds/filemap.cpp

```diff
@@ -1012,134 +1040,195 @@ void FileMapInfo::validate_non_existent_class_paths() {
-    fail_continue("Unable to read the file header.");
-    os::free(dynamic_header);
-    os::close(fd);
+  GenericCDSFileMapHeader* header = file_helper.get_generic_file_header();
+  if (header->_magic != CDS_DYNAMIC_ARCHIVE_MAGIC) {
+    // Not a dynamic header, no need to proceed further.
     return false;
   }
-  if (dynamic_header->magic() != CDS_DYNAMIC_ARCHIVE_MAGIC) {
-    // Not a dynamic header, no need to proceed further.
-    *size = 0;
-    os::free(dynamic_header);
-    os::close(fd);
+
+  if ((header->_base_archive_name_size == 0 && header->_base_archive_path_offset != 0) ||
+      (header->_base_archive_name_size != 0 && header->_base_archive_path_offset == 0)) {
+    fail_continue("Default base archive not set correct");
     return false;
   }
-  if (dynamic_header->base_archive_is_default()) {
+  if (header->_base_archive_name_size == 0 &&
+      header->_base_archive_path_offset == 0) {
     *base_archive_name = Arguments::get_default_shared_archive_path();
   } else {
     // read the base archive name
-    size_t name_size = dynamic_header->base_archive_name_size();
-    if (name_size == 0) {
-      os::free(dynamic_header);
-      os::close(fd);
-      return false;
-    }
-    *base_archive_name = NEW_C_HEAP_ARRAY(char, name_size, mtInternal);
-    n = os::read(fd, *base_archive_name, (unsigned int)name_size);
-    if (n != name_size) {
-      fail_continue("Unable to read the base archive name from the header.");
-      FREE_C_HEAP_ARRAY(char, *base_archive_name);
+    if (!file_helper.read_base_archive_name(base_archive_name)) {
       *base_archive_name = NULL;
-      os::free(dynamic_header);
-      os::close(fd);
       return false;
     }
   }
-
-  os::free(dynamic_header);
-  os::close(fd);
   return true;
 }
 
 // Read the FileMapInfo information from the file.
 
 bool FileMapInfo::init_from_file(int fd) {
-  size_t sz = is_static() ? sizeof(FileMapHeader) : sizeof(DynamicArchiveHeader);
-  size_t n = os::read(fd, header(), (unsigned int)sz);
-  if (n != sz) {
+  FileHeaderHelper file_helper;
+  if (!file_helper.initialize(fd)) {
     fail_continue("Unable to read the file header.");
     return false;
   }
-
-  if (!Arguments::has_jimage()) {
-    FileMapInfo::fail_continue("The shared archive file cannot be used with an exploded module build.");
-    return false;
-  }
+  GenericCDSFileMapHeader* gen_header = file_helper.get_generic_file_header();
 
   unsigned int expected_magic = is_static() ? CDS_ARCHIVE_MAGIC : CDS_DYNAMIC_ARCHIVE_MAGIC;
-  if (header()->magic() != expected_magic) {
+  if (gen_header->_magic != expected_magic) {
     log_info(cds)("_magic expected: 0x%08x", expected_magic);
-    log_info(cds)("         actual: 0x%08x", header()->magic());
+    log_info(cds)("         actual: 0x%08x", gen_header->_magic);
     FileMapInfo::fail_continue("The shared archive file has a bad magic number.");
     return false;
   }
 
+  _header = (FileMapHeader*)os::malloc(gen_header->_header_size, mtInternal);
```

Comment:
There's no need to allocate and read the header again. It's already in gen_header. This should be enough:

```
_header = (FileMapHeader*)gen_header;
```

Reply:
see above reply. We need read the full _header_size for _header. Also note that helper class will delete gen_header when out of scope.